src/app/cluster-building-blocks/QuieterReporting.h

```diff
@@ -0,0 +1,194 @@
+        mLastDirtyTimestampMillis = now;
+        mLastDirtyValue = mValue;
+    }
+
+    /**
+     * Set the updated value of the attribute, computing whether it needs to be reported according to `changedPredicate` and policies.
+     *
+     * - Any change of nullability between `newValue` and the old value will be considered dirty.
+     * - The policies from `QuieterReportingPolicyEnum` and set via `SetPolicy()` are self-explanatory by name.
+     * - The changedPredicate will be called with last dirty <timestamp, value> and new <timestamp value> and may override
+     *   the dirty state altogether when it returns true. Use sparingly and default to a functor returning false.
+     *
+     * Internal recording will be done about last dirty value and last dirty timestamp based on the policies having applied.
+     *
+     * @param newValue - new value to set for the attribute
+     * @param now - system monotonic timestamp at the time of the call
+     * @param changedPredicate - functor to possibly override dirty state
+     */
+    void SetValue(const chip::app::DataModel::Nullable<T>& newValue, Timestamp now, SufficientChangePredicate changedPredicate)
+    {
+        bool isChangeOfNull = newValue.IsNull() ^ mValue.IsNull();
+        bool areBothValuesNonNull = !newValue.IsNull() && !mValue.IsNull();
+
+        bool changeToFromZero = areBothValuesNonNull && (*newValue == 0 || *mValue == 0);
+        bool isIncrement = areBothValuesNonNull && (*newValue > *mValue);
+        bool isDecrement = areBothValuesNonNull && (*newValue < *mValue);
+
+        bool wasDirty = mIsDirty;
+
+        mIsDirty = mIsDirty || isChangeOfNull;
+        mIsDirty = mIsDirty || (mPolicyFlags.Has(QuieterReportingPolicyEnum::kMarkDirtyOnChangeToFromZero) && changeToFromZero);
+        mIsDirty = mIsDirty || (mPolicyFlags.Has(QuieterReportingPolicyEnum::kMarkDirtyOnDecrement) && isDecrement);
+        mIsDirty = mIsDirty || (mPolicyFlags.Has(QuieterReportingPolicyEnum::kMarkDirtyOnIncrement) && isIncrement);
+        mIsDirty = mIsDirty || changedPredicate(mLastDirtyTimestampMillis, now, mLastDirtyValue, newValue);
+
+        mValue = newValue;
+
+        if (!wasDirty && mIsDirty)
+        {
+            mLastDirtyValue = newValue;
+            mLastDirtyTimestampMillis = now;
+        }
+    }
+
+    /**
+     * Same as the other `SetValue`, but assumes a changedPredicate that never overrides to dirty.
+     *
+     * This is the easy/common case.
+     *
+     * @param newValue - new value to set for the attribute
+     * @param now - system monotonic timestamp at the time of the call
+     */
+    void SetValue(const chip::app::DataModel::Nullable<T>& newValue, Timestamp now)
+    {
+        SetValue(newValue, now, [](Timestamp, Timestamp, const Nullable<T> &, const Nullable<T> &) -> bool { return false; });
+    }
+
+  protected:
+    chip::app::DataModel::Nullable<T> mValue;
+    chip::app::DataModel::Nullable<T> mLastDirtyValue;
+    bool mIsDirty = false;
```

Comment:
A blurb on what each of these contain and relation between values (e.g. mLastDirtyTimestampMillis being updated whenever mLastDirtyValue is updated) would help understanding how members are used, and mLastDirtyValue lags behind mValue and is updated whenever mDirty transitions.

Reply:
Use of mLastDirtyValue seems to be `mFirstDirtyValue` as it is used in SetValue: once something is dirty, we set the mIsDirty flag and never update the dirty value ever again. Maybe just naming `dirtyValue` would be better (as the forcing dirty does indeed force the value to be the last).